### DISCUSSION TITLE
Some questions about the provider

### DISCUSSION THREAD

**Discussion by iambinlin** [pmndrs/jotai#2619] [Q&A] [answered]
When using react-router-dom, multiple routes correspond to multiple Providers. If a key is not added, it will lead to abnormal state behavior, meaning the state will not be cleaned up when the page is unmounted. By adding a key, the state of the Provider will automatically reset when the page is unmounted.

eg: 
```javascript
<BrowserRouter>
    <Route><Provider>{pageChildren}</Provider><Route>
    <Route><Provider>{pageChildren}</Provider><Route>
    <Route><Provider>{pageChildren}</Provider><Route>
<BrowserRouter>
```

to

```javascript
<BrowserRouter>
    <Route><Provider key={route.path}>{pageChildren}</Provider><Route>
    <Route><Provider key={route.path}>{pageChildren}</Provider><Route>
    <Route><Provider key={route.path}>{pageChildren}</Provider><Route>
<BrowserRouter>

```

**Comment by iambinlin** [comment]
I just have some doubts and haven't thoroughly investigated.

**Reply by dai-shi** [reply, marked as answer]
Adding key seems the right solution.

**Reply by iambinlin** [reply]
I will try to reproduce it later.

**Comment by iambinlin** [comment]
[Code SanBox Demo](https://codesandbox.io/p/sandbox/jotai-multi-provider-8j7ccp?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxj3du6e00063b6kcoasicmm%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxj3du6d00023b6k8j3yexyp%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxj3du6e00033b6k9hw35qtg%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxj3du6e00053b6k2r094g13%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxj3du6d00023b6k8j3yexyp%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxj3du6d00013b6ke8q1clrf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%257D%255D%252C%2522id%2522%253A%2522clxj3du6d00023b6k8j3yexyp%2522%252C%2522activeTabId%2522%253A%2522clxj3du6d00013b6ke8q1clrf%2522%257D%252C%2522clxj3du6e00053b6k2r094g13%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxj3du6e00043b6kve9q5aeb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2523%252F%2522%257D%255D%252C%2522id%2522%253A%2522clxj3du6e00053b6k2r094g13%2522%252C%2522activeTabId%2522%253A%2522clxj3du6e00043b6kve9q5aeb%2522%257D%252C%2522clxj3du6e00033b6k9hw35qtg%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clxj3du6e00033b6k9hw35qtg%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
Hello,  @dai-shi, I have successfully reproduced this scenario in CodeSandbox. The /app route has a key added, so every time you navigate into it from the nav and then go back, the atom can reset. The /home route does not have a key added, so every time you navigate into it from the nav and then go back, the atom does not get cleared.
<img width="1112" alt="image" src="https://github.com/pmndrs/jotai/assets/39019913/0055b317-d7a4-4289-be8b-530a4a2b4e46">

I'm not sure if this is an issue or if there's a problem with my usage. I would like to ask for your help.

**Reply by dai-shi** [reply]
I assume it's the same behavior as a native React Context.
Jotai Provider is a simple wrapper around React Context. It's only a few lines of code.
So, the question is if react-router-dom works with native React Context without Jotai.

**Reply by iambinlin** [reply]
> I assume it's the same behavior as a native React Context. Jotai Provider is a simple wrapper around React Context. It's only a few lines of code. So, the question is if react-router-dom works with native React Context without Jotai.

Sure, I can confirm this because in the hundreds of projects I've worked on, it's almost always in this form. Let me dive deeper into debugging the state transition behavior.
Thank you for your response. I'll comment here with any updates later.

**Reply by iambinlin** [reply]
![Kapture 2024-06-19 at 00 22 17](https://github.com/pmndrs/jotai/assets/39019913/e2b948e5-b65e-45b1-9274-b4ed88b55bba)

I have added logging for re-renders in the Provider source code (useEffect execution indicates a re-render).

----

1. **Without adding a key**: The page does not trigger re-renders upon each entry, reusing the `App's Provider`.
2. **With a key added**: The page re-renders every time because the different keys bypass React's rendering optimization.

<img width="1024" alt="image" src="https://github.com/pmndrs/jotai/assets/39019913/dec4af8f-14ba-4f47-bda0-a7915d880e73">

--- 
When using the same `Provider` component, it seems that React **reuses it**.
 
<img width="1379" alt="image" src="https://github.com/pmndrs/jotai/assets/39019913/cda94367-8d1f-4235-a471-a3c37b795088">

In other words, the `APP Provider` manages the internal Provider. Unless a key is added manually, React's rendering mechanism will reuse it.

The above steps are my preliminary investigation into this issue. The accuracy of my conclusions is still uncertain. I hope you can help verify my steps and assumptions.

It seems that your initial response was entirely correct; adding a key is the right choice, especially in scenarios with an `APP-level Provider` where the `Page Provider` might only serve as a dynamically added or removed page element.

-------
My goal is to isolate:
For an application, the `APP Provider` manages the global state of the entire application (cleared when the entire site is closed), while the `Page Provider` manages the page-level state (cleared when the page is unloaded). Further down, there is more granular control, such as at the `component level`.

**Reply by iambinlin** [reply]
Thank you again for taking the time to answer my question despite your busy schedule.

**Comment by iambinlin** [comment]
I can also encapsulate a Provider with a key as a common component for reuse.
```typescript
import { useRef } from 'react';
import { Provider } from 'jotai' ;
import { nanoId } from 'nanoid';
export default function WrapperProvider({ children, ...props }) {
    const scopeKey = useRef(nanoId(7)).current;
    return <Provider {...props} key={scopeKey}>{children}<Provider>
}
```

**Comment by dai-shi** [comment]
Also check out https://github.com/jotaijs/jotai-scope which may or may not be related.

**Reply by iambinlin** [reply]
<img width="735" alt="image" src="https://github.com/pmndrs/jotai/assets/39019913/c8101335-e253-41d0-970a-8834b5cafa0d">

It meets the requirements. I took a look, and the core part seems to be dynamically `creating context` within the function.
It's somewhat like a `singleton` or `multi-instance`, but not exactly. Let me take a look, thanks.

It looks like declarative scope collection. I have used recoil before, which automatically collects internally. If it automatically collects the atoms used in child components, is that feasible?

Additionally, it seems that adding a key to the `Provider` can also achieve this effect. Would there be any side effects?

**Reply by dai-shi** [reply]
> If it automatically collects the atoms used in child components, is that feasible?

I don't know how Recoil works, but Jotai doesn't have such a capability and thus it's not feasible to do something in `jotai-scope`. But, it might not be completely impossible.

> Would there be any side effects?

If adding `key` works for you, it's much simpler and less error-prune.

**Reply by iambinlin** [reply]
Get, thank u very much～
